File: frontend/src/expiryUtils.ts
/** Options expire at 5:30 PM IST on the expiry calendar day. */

function parseExpiry(expiry: string): Date {
  const [day, month, year] = expiry.split('-').map(Number);
  // 5:30 PM IST on expiry day = 12:00 UTC on the same calendar date
  return new Date(Date.UTC(year, month - 1, day, 12, 0, 0));
}

export function sortExpiries(expiries: string[]): string[] {
  return [...expiries].sort((a, b) => parseExpiry(a).getTime() - parseExpiry(b).getTime());
}

/** Expiries not yet settled (cutoff 5:30 PM IST on expiry day). */
export function activeExpiries(expiries: string[], nowMs = Date.now()): string[] {
  return sortExpiries(expiries).filter((e) => parseExpiry(e).getTime() > nowMs);
}

export interface ExpirySlotOption {
  value: string;
  label: string;
}

export function expirySlotValue(index: number): string {
  if (index === 0) return 'today';
  if (index === 1) return 'tomorrow';
  return `slot_${index}`;
}

export function expirySlotOptions(expiries: string[]): ExpirySlotOption[] {
  const active = activeExpiries(expiries);
  return active.map((date, i) => {
    const value = expirySlotValue(i);
    if (i === 0) return { value, label: `Today (${date})` };
    if (i === 1) return { value, label: `Tomorrow (${date})` };
    return { value, label: date };
  });
}

export function expirySlotLabel(slot: string, expiries: string[]): string {
  const match = expirySlotOptions(expiries).find((o) => o.value === slot);
  return match?.label ?? slot;
}

export function expirySlotLabels(expiries: string[]): { today?: string; tomorrow?: string } {
  const active = activeExpiries(expiries);
  return {
    today: active[0],
    tomorrow: active[1],
  };
}

File: backend/app/expiry_utils.py
from datetime import datetime
from zoneinfo import ZoneInfo

IST = ZoneInfo("Asia/Kolkata")
EXPIRY_HOUR = 17
EXPIRY_MINUTE = 30


def parse_expiry_date(expiry: str) -> datetime:
    """Parse DD-MM-YYYY expiry to 5:30 PM IST on that calendar day."""
    day, month, year = map(int, expiry.split("-"))
    return datetime(year, month, day, EXPIRY_HOUR, EXPIRY_MINUTE, tzinfo=IST)


def sort_expiries(expiries: list[str]) -> list[str]:
    return sorted(expiries, key=parse_expiry_date)


def active_expiries(expiries: list[str], now: datetime | None = None) -> list[str]:
    """Expiries that have not yet settled (cutoff 5:30 PM IST on expiry day)."""
    now = now or datetime.now(IST)
    return [e for e in sort_expiries(expiries) if parse_expiry_date(e) > now]


def resolve_expiry_slot(slot: str, expiries: list[str]) -> str:
    """
    today = nearest active expiry
    tomorrow = second nearest active expiry
    slot_N = Nth active expiry (slot_2, slot_3, ...)
    """
    active = active_expiries(expiries)
    if not active:
        raise ValueError("No active option expiries available")
    if slot == "today":
        index = 0
    elif slot == "tomorrow":
        index = 1
    elif slot.startswith("slot_"):
        try:
            index = int(slot.split("_", 1)[1])
        except (ValueError, IndexError) as e:
            raise ValueError(f"Invalid expiry slot '{slot}'") from e
    else:
        raise ValueError(f"Invalid expiry slot '{slot}'")
    if index >= len(active):
        raise ValueError(f"Not enough expiries for '{slot}' (only {len(active)} active)")
    return active[index]
